### PR TITLE
fix: improve Embedded Yield transfer feedback

### PR DESCRIPTION
**Pull request by keeganthomp**
## What
- Shows a submitted Embedded Yield movement in Recent activity immediately.
- Keeps the requested withdrawal amount visible until the API returns the exact payout, while live status and API values remain authoritative.
- Replaces pending toast copy with explicit confirmed or failed messaging.

## Why
Dashboard reads can lag the submit response, and withdrawal `tokenAmount` stays null until settlement. The previous fallback was tied to balance projection, so the row could appear late or revert to a dash before finalization.

## Verification
- `pnpm test`: 62 passed
- `pnpm typecheck`
- `pnpm lint`
- `pnpm build`

### CODE REVIEW

**Comment by cutbow7** [comment]
# Review: fix: improve Embedded Yield transfer feedback (#1943)

**Verdict: Approve** — the fix works as described, all checks pass, and nothing I found is merge-blocking. Three non-blocking nits below, each reproduced against the actual code.

## What this does

- `applySubmittedTransfers` (new, `src/lib/movements.ts:98`) keeps this tab's submit responses so Recent activity appears immediately and a withdrawal keeps showing the requested amount until SDP reports the exact payout. The `tokenAmount` fill responsibility moved out of `applyInFlight` (which now only projects balances) — the old behavior is re-covered by the new `submitted transfer activity` test block, so removing the old test is fine.
- Toast copy is centralized in `confirmationDescription` (`src/App.tsx:63`), failed toasts now say "No money was moved.", and the JSX shell is extracted into `BankShell` (`src/App.tsx:394`) with identical render semantics (`loading`/`error`/`view` conditions are behavior-preserving).

## Verification (all run locally, pnpm only)

| Check | Result |
| --- | --- |
| `pnpm install --frozen-lockfile` (pnpm 10.16.0, node 24.15.0) | ✅ |
| `pnpm test` | ✅ 62/62 passed (matches PR claim) |
| `pnpm typecheck` | ✅ clean |
| `pnpm lint` (biome) | ✅ clean |
| `pnpm build` (`next build`) | ✅ clean |
| Production build booted on `127.0.0.1:4173` | ✅ unauthenticated → 401, basic-auth → 200, `/api/dashboard` returns a controlled error JSON with no SDP backend (designed fail-closed path) |
| React Doctor (`--scope changed --base origin/main --project sdp-web,sdp-docs`) | ✅ no diagnostics — the PR only touches `examples/embedded-yield-bank`, which is outside the sdp-web/sdp-docs CI scope, so CI will also see nothing to report |

I additionally wrote throwaway tests against the real source to confirm each edge case below, then deleted them (final `git status` clean, suite re-run green).

## Findings (all non-blocking)

### 1. A failed movement shows the requested amount as if it moved — forever

`transfer()` inserts the submit response into `submittedTransfers` *before* the `status === "failed"` check (`src/App.tsx:308-316`), and `applySubmittedTransfers` fills any `tokenAmount: null` regardless of status (`src/lib/movements.ts:114`, `src/lib/movements.ts:124-128`). A failed withdrawal therefore renders a row with the requested amount (reproduced: `tokenAmount === "1.5"` on a `failed` movement), while the toast says "No money was moved." Since a failed movement never gets an exact payout from the API, this contradicts the "live status and API values remain authoritative" intent.

**Suggested fix** in `applySubmittedTransfers` — skip the fill for failed movements:

```ts
.map(({ movement, requestedTokenAmount }) => ({
  ...movement,
  tokenAmount:
    movement.status === "failed"
      ? movement.tokenAmount
      : (movement.tokenAmount ?? requestedTokenAmount),
}))
```

(and the same guard in the `movement.tokenAmount === null && submittedTransfer` branch at `src/lib/movements.ts:124`).

### 2. `submittedTransfers` is never pruned

Entries are only ever prepended (`src/App.tsx:308-313`); nothing removes them for the lifetime of the tab. Besides unbounded growth (demo-scale, but still), I reproduced that a movement which later **disappears from the dashboard read is resurrected** by `optimisticOnly` (`src/lib/movements.ts:110-116`) — a stale row that can never reconcile if SDP ever drops or filters it server-side.

**Suggested fix**: in `refresh()` after `setData(next)`, drop entries the live read has fully reconciled:

```ts
setSubmittedTransfers((current) =>
  current.filter(
    (t) =>
      !next.movements.some(
        (m) => m.movementId === t.movement.movementId && m.tokenAmount !== null
      )
  )
);
```

### 3. Optimistic rows always sort above newer live rows

`optimisticOnly` is unconditionally prepended (`src/lib/movements.ts:120-121`) after sorting only within itself (`src/lib/movements.ts:116`). If another tab/source created a *newer* movement that is already in the live read, this tab's older optimistic row still renders above it. Reproduced. Fix: concatenate optimistic + live movements and sort the whole array by `createdAt` desc instead of prepending.

## Notes

- `React Doctor` ran clean locally; in my checkout `origin` points at a fork, so I aliased `origin/main` to the fetched upstream `main` (`ac632aaa`) before running — same command CI uses, same (empty) result.
- The removed `applyInFlight` test ("fills a pending withdrawal's amount...") is fully superseded by the new `submitted transfer activity` block — no coverage lost.

File: examples/embedded-yield-bank/src/App.tsx [hunk 1]
```diff
@@ -267,6 +305,12 @@ export function App() {
       const { movement } = await (direction === "to-savings"
         ? createDeposit(amount)
         : createWithdrawal(amount));
+      setSubmittedTransfers((current) => [
+        { movement, requestedTokenAmount: amount },
+        ...current.filter(
+          (transfer) => transfer.movement.movementId !== movement.movementId
+        ),
+      ]);
```
[the remark this answers is not in the document]
Comment (keeganthomp) [reply examples/embedded-yield-bank/src/App.tsx:320]: Addressed in [`faaaf48dfeca`](https://github.com/solana-foundation/solana-developer-platform/commit/faaaf48dfecacb3f9c35964512174d48e5d66800): Pruned overlays after an exact amount or failure reaches the live ledger, and covered both outcomes in the 65-test suite.

File: examples/embedded-yield-bank/src/lib/movements.ts [hunk 2]
```diff
@@ -84,6 +84,54 @@ export interface InFlightTransfer {
...
+      ...movement,
+      tokenAmount: movement.tokenAmount ?? requestedTokenAmount,
+    }))
+    .sort((left, right) => right.createdAt.localeCompare(left.createdAt));
+
+  return {
+    ...live,
+    movements: [
+      ...optimisticOnly,
+      ...live.movements.map((movement) => {
```
[the remark this answers is not in the document]
Comment (keeganthomp) [reply examples/embedded-yield-bank/src/lib/movements.ts]: Addressed in [`faaaf48dfeca`](https://github.com/solana-foundation/solana-developer-platform/commit/faaaf48dfecacb3f9c35964512174d48e5d66800): Merged optimistic and live activity into one createdAt-descending list; the newer-live-row regression is covered and all 65 tests pass.

**Comment by keeganthomp** [comment]
Also addressed the failed-withdrawal nit in `faaaf48dfeca`: known failed submit responses never enter the optimistic overlay, and requested amounts are never filled onto failed rows. Added regression coverage; 65/65 tests, typecheck, lint, and build pass.

**Review by cutbow7 — APPROVED** [review]
